keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+store it once. Also, users are already using CRDs for a similar function, and, for those users, the API Server already 
+tolerates the additional storage burden.
+
+## Alternatives 
+This section contains various design alternatives that were considered during the generation of this KEP.
+
+### Use Labeling Conventions
+Rather than introducing an API object, we could require that all tools use a standard labeling scheme to identify 
+the components of an Application. This would also provide an opt-in method for tool interoperability. However, this 
+method provides no way to garbage collect Application components, express dependencies between applications, or to 
+aggregate application meta-data in a single, well-known object. Also, there is no way to ensure that tools implement 
+labeling conventions consistently or correctly. Well intentioned implementors may accidentally produce tools that 
+incorrectly implement the convention leading to degraded or broken user experiences for other tools or UIs.
+This method lacks the capacity for server side validation of the applications definition. There is not way to 
+ensure a collection of labels correctly communicates the users intentions. Lastly, this method lacks any 
+method to provide health checks.
+
+### Use Custom Resource Definitions
+Rather than introducing an API object, each tool could simply use a CRD (Custom Resource Definition) to represent 
+Applications. As CRDs are, by definition, a custom API extension mechanism, it is unlikely that this method would 
+provide for the desired level of interoperability across tools and UIs. If each tool uses their own CRD, they will a 
+priori not be interoperable, and no UI could hope to surface relevant information for all of them in a uniform manner. 
+If all tools use a common CRD, then we might as well introduce that resource as a well known Kind.
+
+### Application Type Aggregation
+An Application object must be created for each instance of application and not each type of application. It 
+might be useful to provide an object that aggregates Applications by type (e.g. all instances of WordPress that are 
+installed in the system), but this can also be accomplished by querying for all Applications and filtering or 
+aggregating based on the declared `.Spec.Type` field. In the future, if the Application Kind is widely adopted, it would 
+be possible to extend the system with another Kind that provides this aggregation via the API object model. The 
+challenges here would be multi-tenancy and namespace isolation. If a higher level object aggregates all Applications 
+of the same Type, it would have to do so across namespaces. This is not compatible with the existing namespace model 
+for Kubernetes unless we assume that the higher level object is only accessible to cluster administrators with 
+permissions to access all namespaces in the cluster.
```

Comment:
I am starting to have misgivings about calling this an "Application". Really, it's more like a _component group_ or _resource collection_. Its primary function really seems to be capturing a list of disparate but related components, but without giving any real sense of the purposiveness or cohesiveness of this set of things. There's no real information about versions, interoperability, or linkages between components... it's just a list of resources.

Moreover, calling this an "application" may be hiding some really cool usages. For example, one _totally legitimate_ use of this system might be to track (and perform certain bulk operations on) a set of ConfigMap/Secret objects. This, I have noticed, is becoming an ops trend. The ops team will maintain a set of configuration files that other applications share. The dev team, say, has no access to create or modify these, but their apps can attach them as appropriate. Using this type to track that collection of configuration data would be really cool, and solve some real world problems... but we have to be honest that it is not at all _an application_.

So perhaps the right move might be to call this something like a ResourceCollection or something, and at a later point overlay a concept of application on top of it (much the way ReplicaSet and Deployment are related).

Reply:
ResourceCollection or Package is fine. I'm not strongly attached to the name.

Reply:
The term App implies something that can be run.  
The License, AppVersion, and DashboardURL attributes only make sense to me for a thing that has image(s) and runs.
A common UX for Apps will be to display aggregate resource usage for the pods, resource usage, image versions, and so on.   

Although this could be used for non-Podful resource collections, it does dilute the value of the App concept in UX to do so.

Suggest for alpha calling this App, and recommending that it be used for collections that include Pods.

Reply:
@erictune there are some concepts here not naturally tying to ResourceCollection like dependencies but more an application level thing. ResourceCollection in your example can simply be done by using labels.